nova_vm/src/engine/bytecode/executable.rs

```diff
@@ -1313,12 +1313,33 @@ impl CompileEvaluation for ast::StaticMemberExpression<'_> {
             ctx.exe.add_instruction(Instruction::GetValue);
         }
 
+        let maybe_jmp = if self.optional {
+            ctx.exe.add_instruction(Instruction::IsNullOrUndefined);
```

Comment:
**issue:** You shouldn't have removed the LoadCopy. That was necessary.

You can think of it like this: Before this `IsNullOrUndefined` we're at a state where stack has a depth of N which we can decide for our purposes to be 0. The result register has our `foo?.bar` expressions `foo` in it. `IsNullOrUndefined` takes the result register value and replaces it with a true if it was null or undefined, otherwise it puts in false. So, we lose `foo` entirely if we do nothing.

So, the `LoadCopy` you had here originally was perfectly correct. With `LoadCopy` and then `IsNullOrUndefined` done we're at a state where stack has depth 1 (we decided it was 0 before), containing `foo` and result register contains true or false. Now if the result is nullish, we should jump to returning undefined. Otherwise we need to move pop `foo` from the stack back into the result register.

So, I think the jump here should actually be `JumpIfTrue` (IIRC I added that a few weeks ago), and after the jump (so if we did not jump from the nullish check) you should add back in the `Store` instruction you had there originally. (Note: Both the LoadCopy and Store should still be done inside the if of course.)

Then, if we from `maybe_jmp` then we arrive at its "receiving end" with a stack depth of 1 (containing `foo`) and an empty result. We always want to leave the function with a stack depth of 0 (otherwise we're leaking stack memory) and in the was-nullish case with `Value::Undefined` in the result register.

So at the receiving end, after `jump_to_end` (which is there to jump over the was-nullish case handling) you need to add `Store` to pop `foo` from the stack before you do the `StoreConstant`. That way we empty the stack before exiting.